Description: [droidcamsrc] implement camerabin2's preview-image interface
 This patch implement camerabin2's preview-image interface interface,
 which make the element post a preview-image bus message when
 post-previews property is enabled.
 .
 As Android doesn't seem to have the concept of preview-image (there are
 raw & postview callbacks but they seem not to be called), I replicate
 what Qt does with it's native Android plugin, which is storing the
 latest viewfinder frame and using that as the preview. This require
 changing how the preview callback flags is calculated as otherwise
 preview callback won't be called. The last preview buffer is kept
 within the structure of dev but use the seperated lock and condition
 variable so that I don't need to think about locking in other places
 (this may change in the future).
 .
 Camerabin2's preview-image interface also require that the element has
 the preview-caps property and must send the preview-image according to
 this caps. To fulfill that, a camerabinpipeline is constructed inside
 the gstdroidcamsrc element. This pipeline will take care of all
 conversion requied to make the preview image comply to the provided
 caps and will also actually posting it to the bus.
 .
 When the image capture is requested, the post-preview property is
 copied to image capturing structure so that the change after the
 capture won't affect this capture. When the shutter callback is
 called and if the preview image is requested, the kept viewfinder
 buffer is sent to the preview pipeline for processing & posting.
 .
 This is required because Qt's camerabin plugin expects the
 preview-image message for its imageExposed and imageCaptured signal
 and some application expect imageCaptured to be fired.
Author: Ratchanan Srirattanamet <peathot@hotmail.com>
Forwarded: https://github.com/sailfishos/gst-droid/pull/36
Last-Update: 2019-08-29
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
diff --git a/gst/droidcamsrc/gstdroidcamsrc.c b/gst/droidcamsrc/gstdroidcamsrc.c
index 385f64e..f5452bc 100644
--- a/gst/droidcamsrc/gstdroidcamsrc.c
+++ b/gst/droidcamsrc/gstdroidcamsrc.c
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2015-2016 Jolla LTD.
+ * Copyright (C) 2010 Texas Instruments, Inc
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -116,6 +117,7 @@ static guint droidcamsrc_signals[LAST_SIGNAL];
 #define DEFAULT_SENSOR_ORIENTATION     0
 #define DEFAULT_IMAGE_MODE             GST_DROIDCAMSRC_IMAGE_MODE_NORMAL
 #define DEFAULT_TARGET_BITRATE         12000000
+#define DEFAULT_POST_PREVIEW           FALSE
 
 static GstDroidCamSrcPad *
 gst_droidcamsrc_create_pad (GstDroidCamSrc * src,
@@ -205,6 +207,11 @@ gst_droidcamsrc_init (GstDroidCamSrc * src)
   src->video = gst_droidcamsrc_mode_new_video (src);
   src->active_mode = NULL;
 
+  src->post_preview = DEFAULT_POST_PREVIEW;
+  src->preview_caps = NULL;
+  src->preview_pipeline =
+      gst_camerabin_create_preview_pipeline (GST_ELEMENT_CAST (src), NULL);
+
   GST_OBJECT_FLAG_SET (src, GST_ELEMENT_FLAG_SOURCE);
 }
 
@@ -302,6 +309,15 @@ gst_droidcamsrc_get_property (GObject * object, guint prop_id, GValue * value,
       g_value_set_int (value, src->target_bitrate);
       break;
 
+    case PROP_POST_PREVIEW:
+      g_value_set_boolean (value, src->post_preview);
+      break;
+
+    case PROP_PREVIEW_CAPS:
+      if (src->preview_caps)
+        gst_value_set_caps (value, src->preview_caps);
+      break;
+
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -402,6 +418,42 @@ gst_droidcamsrc_set_property (GObject * object, guint prop_id,
       src->target_bitrate = g_value_get_int (value);
       break;
 
+    case PROP_POST_PREVIEW:
+      src->post_preview = g_value_get_boolean (value);
+
+      if (src->dev) {
+        gst_droidcamsrc_dev_update_preview_callback_flag (src->dev);
+      }
+
+      break;
+
+    case PROP_PREVIEW_CAPS:{
+      GstCaps *new_caps;
+
+      new_caps = (GstCaps *) gst_value_get_caps (value);
+      if (new_caps == NULL) {
+        new_caps = gst_caps_new_any ();
+      } else {
+        new_caps = gst_caps_ref (new_caps);
+      }
+
+      if (!gst_caps_is_equal (src->preview_caps, new_caps)) {
+        gst_caps_replace (&src->preview_caps, new_caps);
+
+        if (src->preview_pipeline) {
+          GST_DEBUG_OBJECT (src,
+              "Setting preview pipeline caps %" GST_PTR_FORMAT,
+              src->preview_caps);
+          gst_camerabin_preview_set_caps (src->preview_pipeline,
+              src->preview_caps);
+        }
+      } else {
+        GST_DEBUG_OBJECT (src, "New preview caps equal current preview caps");
+      }
+      gst_caps_unref (new_caps);
+    }
+      break;
+
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -432,6 +484,15 @@ gst_droidcamsrc_finalize (GObject * object)
 
   g_rec_mutex_clear (&src->dev_lock);
 
+  if (src->preview_pipeline) {
+    gst_camerabin_destroy_preview_pipeline (src->preview_pipeline);
+    src->preview_pipeline = NULL;
+  }
+
+  if (src->preview_caps) {
+    gst_caps_replace (&src->preview_caps, NULL);
+  }
+
   G_OBJECT_CLASS (parent_class)->finalize (object);
 }
 
@@ -546,6 +607,19 @@ gst_droidcamsrc_change_state (GstElement * element, GstStateChange transition)
         break;
       }
 
+      if (src->preview_pipeline == NULL) {
+        /* failed to create preview pipeline, fail state change */
+        ret = GST_STATE_CHANGE_FAILURE;
+      }
+
+      if (src->preview_caps) {
+        GST_DEBUG_OBJECT (src,
+            "Setting preview pipeline caps %" GST_PTR_FORMAT,
+            src->preview_caps);
+        gst_camerabin_preview_set_caps (src->preview_pipeline,
+            src->preview_caps);
+      }
+
       src->dev =
           gst_droidcamsrc_dev_new (src->vfsrc, src->imgsrc,
           src->vidsrc, &src->dev_lock);
@@ -608,6 +682,11 @@ gst_droidcamsrc_change_state (GstElement * element, GstStateChange transition)
 
       /* Now add the needed orientation tag */
       gst_droidcamsrc_add_vfsrc_orientation_tag (src);
+
+      /* without this the preview pipeline will not post buffer
+       * messages on the pipeline */
+      gst_element_set_state (src->preview_pipeline->pipeline,
+          GST_STATE_PLAYING);
     }
 
       break;
@@ -671,11 +750,15 @@ gst_droidcamsrc_change_state (GstElement * element, GstStateChange transition)
     case GST_STATE_CHANGE_PAUSED_TO_READY:
       gst_droidcamsrc_dev_deinit (src->dev);
       gst_droidcamsrc_dev_close (src->dev);
+
+      gst_element_set_state (src->preview_pipeline->pipeline, GST_STATE_READY);
       break;
 
     case GST_STATE_CHANGE_READY_TO_NULL:
       gst_droidcamsrc_dev_destroy (src->dev);
       src->dev = NULL;
+
+      gst_element_set_state (src->preview_pipeline->pipeline, GST_STATE_NULL);
       break;
 
     default:
@@ -1082,6 +1165,17 @@ gst_droidcamsrc_class_init (GstDroidCamSrcClass * klass)
           "Supported ISO speeds", G_VARIANT_TYPE_VARIANT, NULL,
           G_PARAM_READABLE));
 
+  /* camerabin interface */
+  g_object_class_install_property (gobject_class, PROP_POST_PREVIEW,
+      g_param_spec_boolean ("post-previews", "Post Previews",
+          "If capture preview images should be posted to the bus",
+          DEFAULT_POST_PREVIEW, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_PREVIEW_CAPS,
+      g_param_spec_boxed ("preview-caps", "Preview caps",
+          "The caps of the preview image to be posted (NULL means ANY)",
+          GST_TYPE_CAPS, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
   gst_droidcamsrc_photography_add_overrides (gobject_class);
 
   /* Signals */
@@ -2484,3 +2578,14 @@ gst_droidcamsrc_get_video_caps_locked (GstDroidCamSrc * src)
 
   return caps;
 }
+
+void
+gst_droidcamsrc_post_preview (GstDroidCamSrc * src, GstSample * sample)
+{
+  if (src->post_preview) {
+    gst_camerabin_preview_pipeline_post (src->preview_pipeline, sample);
+  } else {
+    GST_DEBUG_OBJECT (src, "Previews not enabled, not posting");
+    gst_sample_unref (sample);
+  }
+}
diff --git a/gst/droidcamsrc/gstdroidcamsrc.h b/gst/droidcamsrc/gstdroidcamsrc.h
index 2504e91..438529a 100644
--- a/gst/droidcamsrc/gstdroidcamsrc.h
+++ b/gst/droidcamsrc/gstdroidcamsrc.h
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2016 Jolla LTD.
+ * Copyright (C) 2010 Texas Instruments, Inc
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -122,6 +123,11 @@ struct _GstDroidCamSrc
 
   gint32 target_bitrate;
 
+  /* camerabin interface */
+  gboolean post_preview;
+  GstCaps *preview_caps;
+  GstCameraBinPreviewPipelineData *preview_pipeline;
+
   /* protected with OBJECT_LOCK */
   gint width;
   gint height;
@@ -147,6 +153,8 @@ gboolean gst_droidcamsrc_apply_params (GstDroidCamSrc * src);
 void gst_droidcamsrc_apply_mode_settings (GstDroidCamSrc * src, GstDroidCamSrcApplyType type);
 void gst_droidcamsrc_update_max_zoom (GstDroidCamSrc * src);
 
+void gst_droidcamsrc_post_preview (GstDroidCamSrc * src, GstSample * sample);
+
 G_END_DECLS
 
 #endif /* __GST_DROIDCAMSRC_H__ */
diff --git a/gst/droidcamsrc/gstdroidcamsrcdev.c b/gst/droidcamsrc/gstdroidcamsrcdev.c
index a53fb36..e7a88f7 100644
--- a/gst/droidcamsrc/gstdroidcamsrcdev.c
+++ b/gst/droidcamsrc/gstdroidcamsrcdev.c
@@ -3,6 +3,8 @@
  *
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2015-2016 Jolla LTD.
+ * Copyright (C) 2020 UBports foundation.
+ * Author: Ratchanan Srirattanamet <ratchanan@ubports.com>
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -50,6 +52,7 @@ struct _GstDroidCamSrcImageCaptureState
 {
   gboolean image_preview_sent;
   gboolean image_start_sent;
+  gboolean preview_image_requested;
 };
 
 struct _GstDroidCamSrcVideoCaptureState
@@ -81,6 +84,7 @@ static gboolean
 gst_droidcamsrc_dev_start_video_recording_raw_locked (GstDroidCamSrcDev * dev);
 static void gst_droidcamsrc_dev_queue_video_buffer_locked (GstDroidCamSrcDev *
     dev, GstBuffer * buffer);
+static void gst_droidcamsrc_dev_post_preview (GstDroidCamSrcDev * dev);
 
 static void
 gst_droidcamsrc_dev_shutter_callback (void *user)
@@ -98,6 +102,11 @@ gst_droidcamsrc_dev_shutter_callback (void *user)
     dev->img->image_start_sent = TRUE;
   }
 
+  if (dev->img->preview_image_requested) {
+    gst_droidcamsrc_dev_post_preview (dev);
+    dev->img->preview_image_requested = FALSE;
+  }
+
   g_rec_mutex_unlock (dev->lock);
 }
 
@@ -280,16 +289,6 @@ gst_droidcamsrc_dev_preview_frame_callback (void *user,
 
   GST_DEBUG_OBJECT (src, "dev preview frame callback");
 
-  /* We are accessing this without a lock because:
-   * 1) We should not be called while preview is stopped and this is when we manipulate this flag
-   * 2) We can get called when we start the preview and we will deadlock because the lock is already held
-   */
-  if (!dev->use_raw_data) {
-    GST_WARNING_OBJECT (src,
-        "preview frame callback called while not when we do not expect it");
-    return;
-  }
-
   buffer = gst_buffer_new_allocate (NULL, mem->size, NULL);
   gst_buffer_fill (buffer, 0, mem->data, mem->size);
 
@@ -303,10 +302,23 @@ gst_droidcamsrc_dev_preview_frame_callback (void *user,
 
   gst_droidcamsrc_dev_prepare_buffer (dev, buffer, rect, &video_info);
 
-  g_mutex_lock (&pad->lock);
-  g_queue_push_tail (pad->queue, buffer);
-  g_cond_signal (&pad->cond);
-  g_mutex_unlock (&pad->lock);
+  g_mutex_lock (&dev->last_preview_buffer_lock);
+  gst_buffer_replace (&dev->last_preview_buffer, buffer);
+  g_cond_signal (&dev->last_preview_buffer_cond);
+  g_mutex_unlock (&dev->last_preview_buffer_lock);
+
+  /* We are accessing this without a lock because:
+   * 1) We should not be called while preview is stopped and this is when we manipulate this flag
+   * 2) We can get called when we start the preview and we will deadlock because the lock is already held
+   */
+  if (dev->use_raw_data) {
+    g_mutex_lock (&pad->lock);
+    g_queue_push_tail (pad->queue, buffer);
+    g_cond_signal (&pad->cond);
+    g_mutex_unlock (&pad->lock);
+  } else {
+    gst_buffer_unref (buffer);
+  }
 }
 
 static void
@@ -535,6 +547,11 @@ gst_droidcamsrc_dev_new (GstDroidCamSrcPad * vfsrc,
   dev->lock = lock;
 
   dev->pool = NULL;
+
+  dev->last_preview_buffer = NULL;
+  g_mutex_init (&dev->last_preview_buffer_lock);
+  g_cond_init (&dev->last_preview_buffer_cond);
+
   dev->use_recorder = FALSE;
   dev->recorder = gst_droidcamsrc_recorder_create (vidsrc);
 
@@ -646,9 +663,14 @@ gst_droidcamsrc_dev_destroy (GstDroidCamSrcDev * dev)
 
   gst_droidcamsrc_recorder_destroy (dev->recorder);
 
+  gst_buffer_replace (&dev->last_preview_buffer, NULL);
+  g_mutex_clear (&dev->last_preview_buffer_lock);
+  g_cond_clear (&dev->last_preview_buffer_cond);
+
   g_slice_free (GstDroidCamSrcImageCaptureState, dev->img);
   g_slice_free (GstDroidCamSrcVideoCaptureState, dev->vid);
   g_slice_free (GstDroidCamSrcDev, dev);
+
   dev = NULL;
 }
 
@@ -738,16 +760,6 @@ gst_droidcamsrc_dev_start (GstDroidCamSrcDev * dev, gboolean apply_settings)
     goto out;
   }
 
-  if (dev->use_raw_data) {
-    GST_INFO_OBJECT (src, "Using raw data mode");
-    droid_media_camera_set_preview_callback_flags (dev->cam,
-        dev->c.CAMERA_FRAME_CALLBACK_FLAG_CAMERA);
-  } else {
-    GST_INFO_OBJECT (src, "Using native buffers mode");
-    droid_media_camera_set_preview_callback_flags (dev->cam,
-        dev->c.CAMERA_FRAME_CALLBACK_FLAG_NOOP);
-  }
-
   if (!droid_media_camera_start_preview (dev->cam)) {
     GST_ERROR_OBJECT (src, "error starting preview");
     goto out;
@@ -755,6 +767,9 @@ gst_droidcamsrc_dev_start (GstDroidCamSrcDev * dev, gboolean apply_settings)
 
   dev->running = TRUE;
 
+  /* Flag update is done here because the function checks for dev->running. */
+  gst_droidcamsrc_dev_update_preview_callback_flag (dev);
+
   ret = TRUE;
 
 out:
@@ -835,17 +850,43 @@ out:
 gboolean
 gst_droidcamsrc_dev_capture_image (GstDroidCamSrcDev * dev)
 {
+  GstDroidCamSrc *src = GST_DROIDCAMSRC (GST_PAD_PARENT (dev->imgsrc->pad));
+
   gboolean ret = FALSE;
   int msg_type = dev->c.CAMERA_MSG_SHUTTER | dev->c.CAMERA_MSG_RAW_IMAGE
       | dev->c.CAMERA_MSG_POSTVIEW_FRAME | dev->c.CAMERA_MSG_COMPRESSED_IMAGE;
 
   GST_DEBUG ("dev capture image");
 
+  if (src->post_preview) {
+    /*
+     * We must ensure that at least 1 preview buffer exists before proceed.
+     * After take_picture() is called, we might not get additional buffer.
+     */
+
+    g_mutex_lock (&dev->last_preview_buffer_lock);
+
+    gint64 end_time = g_get_monotonic_time () + (1 * G_TIME_SPAN_SECOND);
+    while (!dev->last_preview_buffer) {
+      if (!g_cond_wait_until (&dev->last_preview_buffer_cond,
+              &dev->last_preview_buffer_lock, end_time)) {
+        GST_ERROR
+            ("dev post_preview requested but no preview buffer available.");
+        g_mutex_unlock (&dev->last_preview_buffer_lock);
+        return FALSE;           /* Because dev->lock has not been held yet. */
+      }
+    }
+
+    g_mutex_unlock (&dev->last_preview_buffer_lock);
+  }
+
   g_rec_mutex_lock (dev->lock);
 
   dev->img->image_preview_sent = FALSE;
   dev->img->image_start_sent = FALSE;
 
+  dev->img->preview_image_requested = src->post_preview;
+
   if (!droid_media_camera_take_picture (dev->cam, msg_type)) {
     GST_ERROR ("error capturing image");
     goto out;
@@ -866,6 +907,28 @@ gst_droidcamsrc_dev_start_video_recording (GstDroidCamSrcDev * dev)
 
   GST_DEBUG ("dev start video recording");
 
+  if (src->post_preview) {
+    /*
+     * We must ensure that at least 1 preview buffer exists before proceed.
+     * After recording is started, we will not get additional buffer.
+     */
+
+    g_mutex_lock (&dev->last_preview_buffer_lock);
+
+    gint64 end_time = g_get_monotonic_time () + (1 * G_TIME_SPAN_SECOND);
+    while (!dev->last_preview_buffer) {
+      if (!g_cond_wait_until (&dev->last_preview_buffer_cond,
+              &dev->last_preview_buffer_lock, end_time)) {
+        GST_ERROR
+            ("dev post_preview requested but no preview buffer available.");
+        g_mutex_unlock (&dev->last_preview_buffer_lock);
+        return FALSE;
+      }
+    }
+
+    g_mutex_unlock (&dev->last_preview_buffer_lock);
+  }
+
   gst_buffer_pool_set_flushing (dev->pool, TRUE);
 
   g_mutex_lock (&dev->vidsrc->lock);
@@ -895,6 +958,10 @@ gst_droidcamsrc_dev_start_video_recording (GstDroidCamSrcDev * dev)
 
   ret = TRUE;
 
+  /* Send the preview image out if requested. */
+  if (src->post_preview)
+    gst_droidcamsrc_dev_post_preview (dev);
+
 out:
   g_rec_mutex_unlock (dev->lock);
 
@@ -1237,3 +1304,72 @@ gst_droidcamsrc_dev_queue_video_buffer_locked (GstDroidCamSrcDev * dev,
   /* in case stop_video_recording() is waiting for us */
   g_cond_signal (&dev->vid->cond);
 }
+
+void
+gst_droidcamsrc_dev_update_preview_callback_flag (GstDroidCamSrcDev * dev)
+{
+  GstDroidCamSrc *src = GST_DROIDCAMSRC (GST_PAD_PARENT (dev->imgsrc->pad));
+
+  gboolean use_preview_callback;
+
+  g_rec_mutex_lock (dev->lock);
+
+  if (!dev->running) {
+    GST_INFO_OBJECT (src, "preview is not running, defering flag update");
+    goto out;
+  }
+
+  if (dev->use_raw_data) {
+    GST_INFO_OBJECT (src, "preview use raw data mode");
+    use_preview_callback = TRUE;
+  } else if (src->post_preview) {
+    GST_INFO_OBJECT (src, "post_preview enabled, preview buffer required");
+    use_preview_callback = TRUE;
+  } else {
+    GST_INFO_OBJECT (src, "preview callback disabled");
+    use_preview_callback = FALSE;
+  }
+
+  if (use_preview_callback) {
+    droid_media_camera_set_preview_callback_flags (dev->cam,
+        dev->c.CAMERA_FRAME_CALLBACK_FLAG_CAMERA);
+  } else {
+    droid_media_camera_set_preview_callback_flags (dev->cam,
+        dev->c.CAMERA_FRAME_CALLBACK_FLAG_NOOP);
+  }
+
+out:
+  g_rec_mutex_unlock (dev->lock);
+}
+
+static void
+gst_droidcamsrc_dev_post_preview (GstDroidCamSrcDev * dev)
+{
+  GstDroidCamSrc *src = GST_DROIDCAMSRC (GST_PAD_PARENT (dev->imgsrc->pad));
+
+  GST_DEBUG ("post preview image from last viewfinder buffer");
+
+  /*
+   * Because we've ensured that dev->last_preview_buffer exists before we
+   * start and we never remove it, we can be sure that it should exists here
+   * too.
+   */
+  g_mutex_lock (&dev->last_preview_buffer_lock);
+  GstBuffer *buffer = gst_buffer_ref (dev->last_preview_buffer);
+  g_mutex_unlock (&dev->last_preview_buffer_lock);
+
+  GstVideoMeta *video_meta = gst_buffer_get_video_meta (buffer);
+  g_assert (video_meta != NULL);        // Because we added it in _prepare_buffer
+
+  GstVideoInfo video_info;
+  gst_video_info_set_format (&video_info, video_meta->format,
+      video_meta->width, video_meta->height);
+  GstCaps *caps = gst_video_info_to_caps (&video_info);
+
+  GstSample *sample = gst_sample_new (buffer, caps, NULL, NULL);
+
+  gst_buffer_unref (buffer);
+  gst_caps_unref (caps);
+
+  gst_droidcamsrc_post_preview (src, sample);
+}
diff --git a/gst/droidcamsrc/gstdroidcamsrcdev.h b/gst/droidcamsrc/gstdroidcamsrcdev.h
index 62fff33..1ceb6c3 100644
--- a/gst/droidcamsrc/gstdroidcamsrcdev.h
+++ b/gst/droidcamsrc/gstdroidcamsrcdev.h
@@ -3,6 +3,8 @@
  *
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2015-2016 Jolla LTD.
+ * Copyright (C) 2020 UBports foundation.
+ * Author: Ratchanan Srirattanamet <ratchanan@ubports.com>
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -56,6 +58,10 @@ struct _GstDroidCamSrcDev
   DroidMediaCameraConstants c;
   GstVideoFormat viewfinder_format;
 
+  GstBuffer *last_preview_buffer;
+  GMutex last_preview_buffer_lock;
+  GCond last_preview_buffer_cond;
+
   gboolean use_recorder;
   GstDroidCamSrcRecorder *recorder;
 };
@@ -95,6 +101,8 @@ gboolean gst_droidcamsrc_dev_is_running (GstDroidCamSrcDev * dev);
 
 void gst_droidcamsrc_dev_queue_video_buffer (GstDroidCamSrcDev * dev, GstBuffer * buffer);
 
+void gst_droidcamsrc_dev_update_preview_callback_flag (GstDroidCamSrcDev * dev);
+
 G_END_DECLS
 
 #endif /* __GST_DROIDCAMSRC_DEV_H__ */
diff --git a/gst/droidcamsrc/gstdroidcamsrcphotography.h b/gst/droidcamsrc/gstdroidcamsrcphotography.h
index e8f6274..16b1392 100644
--- a/gst/droidcamsrc/gstdroidcamsrcphotography.h
+++ b/gst/droidcamsrc/gstdroidcamsrcphotography.h
@@ -76,7 +76,11 @@ typedef enum
   PROP_MIN_EXPOSURE_TIME,
   PROP_MAX_EXPOSURE_TIME,
   PROP_NOISE_REDUCTION,
-  PROP_EXPOSURE_MODE
+  PROP_EXPOSURE_MODE,
+
+  /* camerabin interface */
+  PROP_POST_PREVIEW,
+  PROP_PREVIEW_CAPS
 } GstDroidCamSrcProperties;
 
 void gst_droidcamsrc_photography_register (gpointer g_iface,  gpointer iface_data);
-- 
2.17.1

